src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,437 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "kepler.h"
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+
+using Magnum::Math::cross;
+using Magnum::Math::dot;
+
+namespace {
+
+// PI isn't standard, so define it here. Should it be added to constants.h?
+static constexpr double PI = Magnum::Math::Constants<double>::pi();
+
+/**
+ * @brief Wrap an angle to the range [0, 2π)
+ */
+static double wrap_angle(double const angle)
+{
+    double wrapped = fmod(angle, 2.0 * PI);
+    if (wrapped < 0)
+    {
+        wrapped += 2.0 * PI;
+    }
+    return wrapped;
+}
+
+/**
+ * @brief Solve Kepler's equation for elliptic orbits (E = M + e * sin(E))
+ * 
+ * @return Eccentric anomaly E
+ */
+static double solve_kepler_elliptic(double const meanAnomaly, double const eccentricity, int steps)
+{
+    double const tolerance = 1.0E-10;
+
+    // If eccentricity is low, then this is a circular orbit, so M = E
+    if (std::abs(eccentricity) < 1.0E-9)
+    {
+        return meanAnomaly;
+    }
+
+    double delta;
+
+    // initial guess
+    double E = meanAnomaly;
+
+    // use newton's method
+    // todo: implement something which converges more quickly
+    do
+    {
+        double f_E = E - eccentricity * sin(E) - meanAnomaly;
+        double df_E = 1.0 - eccentricity * cos(E);
+
+        delta = f_E / df_E;
+        E = E - delta;
+
+        E = wrap_angle(E); // Sometimes this sometimes fails to converge without this line
+        --steps;
+    } while (std::abs(delta) > tolerance && steps > 1);
+    return E;
+}
+
+/**
+ * @brief Solve Kepler's equation for hyperbolic orbits (E = M + e * sinh(E))
+ * 
+ * @return Eccentric anomaly E
+ */
+static double solve_kepler_hyperbolic(double const meanAnomaly, double const eccentricity, int steps)
+{
+    double const tolerance = 1.0E-10;
+
+    double delta;
+
+    // initial guess
+    double E = std::log(std::abs(meanAnomaly));
+
+    // use newton's method
+    do
+    {
+        double f_E = eccentricity * sinh(E) - E - meanAnomaly;
+        double df_E = eccentricity * cosh(E) - 1.0;
+
+        delta = f_E / df_E;
+        E = E - delta;
+        --steps;
+    } while (std::abs(delta) > tolerance && steps > 1);
+    return E;
+}
+
+static double elliptic_eccentric_anomaly_to_true_anomaly(double const &ecc, double const &E)
```

Comment:
can you add some doxygen documentation?

don't go overboard. Just a one sentence explanation of the point of the function is enough.


extrapolate to all the functions :-)